send-transaction-service/src/transaction_client.rs

```diff
@@ -115,13 +140,143 @@ where
+    leader_info_provider: CurrentLeaderInfo<T>,
+    my_tpu_address: SocketAddr,
+    tpu_peers: Option<Vec<SocketAddr>>,
+}
+
+#[async_trait]
+impl<T> LeaderUpdater for SendTransactionServiceLeaderUpdater<T>
+where
+    T: TpuInfoWithSendStatic,
+{
+    fn next_leaders(&mut self, lookahead_leaders: usize) -> Vec<SocketAddr> {
+        let discovered_peers = self
+            .leader_info_provider
+            .get_leader_info()
+            .map(|leader_info| {
+                leader_info.get_leader_tpus(lookahead_leaders as u64, Protocol::QUIC)
+            })
+            .filter(|addresses| !addresses.is_empty())
+            .unwrap_or_else(|| vec![&self.my_tpu_address]);
+        let mut all_peers = self.tpu_peers.clone().unwrap_or_default();
+        all_peers.extend(discovered_peers.into_iter().cloned());
+        all_peers
+    }
+    async fn stop(&mut self) {}
+}
+
+#[derive(Clone)]
+pub struct TpuClientNextClient {
+    runtime_handle: Handle,
+    sender: Sender<TransactionBatch>,
+    cancel: CancellationToken,
+}
+
+impl Cancelable for TpuClientNextClient {
+    fn cancel(&self) {
+        self.cancel.cancel();
+    }
+}
+
+impl TransactionClient for TpuClientNextClient {
+    fn send_transactions_in_batch(
+        &self,
+        wire_transactions: Vec<Vec<u8>>,
+        stats: &SendTransactionServiceStats,
+    ) {
+        let mut measure = Measure::start("send-us");
+        self.runtime_handle.spawn({
+            let sender = self.sender.clone();
+            async move {
+                let res = sender.send(TransactionBatch::new(wire_transactions)).await;
+                if res.is_err() {
+                    warn!("Failed to send transaction to channel: it is closed.");
+                }
+            }
+        });
+
+        measure.stop();
+        stats.send_us.fetch_add(measure.as_us(), Ordering::Relaxed);
+        stats.send_attempt_count.fetch_add(1, Ordering::Relaxed);
+    }
+
+    fn protocol(&self) -> Protocol {
+        Protocol::QUIC
+    }
+}
+
+pub fn spawn_tpu_client_send_txs<T>(
+    runtime_handle: Handle,
+    my_tpu_address: SocketAddr,
+    tpu_peers: Option<Vec<SocketAddr>>,
+    leader_info: Option<T>,
+    leader_forward_count: u64,
+    validator_identity: Option<Keypair>,
+) -> TpuClientNextClient
+where
+    T: TpuInfoWithSendStatic,
+{
+    let leader_info_provider = CurrentLeaderInfo::new(leader_info);
+
+    let (sender, receiver) = mpsc::channel(128);
+    let cancel = CancellationToken::new();
+    let _handle = runtime_handle.spawn({
+        let cancel = cancel.clone();
+        async move {
+            let leader_updater: SendTransactionServiceLeaderUpdater<T> =
+                SendTransactionServiceLeaderUpdater {
+                    leader_info_provider,
+                    my_tpu_address,
+                    tpu_peers,
+                };
+            let config = ConnectionWorkersSchedulerConfig {
+                bind: SocketAddr::new(Ipv4Addr::new(0, 0, 0, 0).into(), 0),
```

Comment:
wasn't the bind address configurable somewhere?

Reply:
You are right that there is `--bind-address` validator argument [here](https://github.com/anza-xyz/agave/blob/master/validator/src/main.rs#L1108 ). But it is not used in client code (not sure if this is by design):

```rust
 let client_socket = solana_net_utils::bind_in_range(
                IpAddr::V4(Ipv4Addr::UNSPECIFIED),
                VALIDATOR_PORT_RANGE,
            )
```
See https://github.com/anza-xyz/agave/blob/master/quic-client/src/nonblocking/quic_client.rs#L138

Probably, I should create socket using the same function `bind_in_range` here. Will ask on discord if there is any sense to use `UNSPECIFIED` address even when user specified address.

Reply:
I changed this code to use `bind_in_range` instead of `0.0.0.0:0`. There is no sense to introduce `bind` as argument because this would require to do some changes on the validator code to propagate `bind_address` from validator/src/main.rs down to the current client code, which seems to be a bit invasive for this PR.

Reply:
After some discussions about this bind_in_range with Alex, I came to conclusion that there is no need in using bind_in_range and it is always better to use ephemeral client ports.

Regarding using `bind_address` here: it is possible but requires plumbing the whole call-stack from validator/src/main.rs. Looks invasive for this PR.